Purchases/SwiftSources/Purchasing/Transaction.swift

```diff
@@ -0,0 +1,32 @@
+//
+//  Transaction.swift
+//  Purchases
+//
+//  Created by RevenueCat.
+//  Copyright © 2020 Purchases. All rights reserved.
+//
+
+import Foundation
+
+@objc(RCTransaction) public class Transaction: NSObject {
```

Comment:
all objects that inherit from `NSObject` have a public `init` with no arguments. 
If we want to prevent that from being used, maybe we can do something like
```swift
required init?() { fatalError("init() has not been implemented") }
```

Reply:
haven't tested that, though

Reply:
to provide some context on the _why_: 
there's a bit of a gotcha when it comes to swift-objc interoperability and initialization, and nullability. 
if you call init() from objc, that will work, because this is an NSObject. however, the object's properties will not have been initialized, so revenuecatId for example will be nil. 
but since the type is declared as non-nil, you won't be able to check against that - if you do 
```swift
if revenuecatId != nil {
```
the compiler will complain because the object isn't nullable. so you'll be in a bad place. 

Here's a blog post that goes into a bit more detail in case you're curious: 
https://fabiancanas.com/blog/2020/1/9/swift-undefined-behavior.html

Reply:
Thanks for the explanation! Unfortunately it doesn't work, it says "Failable initializer 'init()' cannot override a non-failable initializer". Since it throws a `fatalError` this is actually not a Failable Initializer right? I think it should be this instead:

```
    required override init() { fatalError("init() has not been implemented") }
```

Reply:
I missed the override, but yeah, it should be there. 
does failable mean that it's an `init?()` vs `init()`?

Reply:
yeah failable means `init?`, but since this is throwing a fatalError, it is not an `init?` because it will actually never finish initializing. If I understand correctly, to make it failable it would have to return nil instead of throwing the error, it needs to return something

Reply:
yeah, for a failable it'd be best practice to return nil if it fails. 
I'm good with anything, as long as we make sure that `[[Transaction alloc] init];` can't be used

Reply:
maybe [this](https://stackoverflow.com/a/49159408/5074358) is the way to go?